### ISSUE TITLE
`Repr` instances for basic types

### ISSUE THREAD

**Issue by Vtec234** [leanprover/lean4#619]
For many basic types such as `Expr`, `Level`, `Format` etc, `Repr` cannot be derived at the definition site because of bootstraping issues. However having these would be exceedingly useful for debugging. Should we for example have an `Instances` file somewhere down from `Repr` in the dependency hierarchy which defines these?

**Comment by leodemoura** [comment]
The instances for `Expr` and `Level` can be synthesized when they are declared. There is no bootstrapping issue. However, I am not sure how useful the automatically generated ones are because they include the `Data` fields.
Here is an example
```lean
Lean.Expr.lam
  `α
  (Lean.Expr.sort (Lean.Level.succ (Lean.Level.zero 2221) 1101187905297) 1469443585)
  (Lean.Expr.lam
    `x
    (Lean.Expr.const `Nat [] 607656011)
    (Lean.Expr.lam
      `y
      (Lean.Expr.bvar 1 2203303673837)
      (Lean.Expr.app
        (Lean.Expr.app
          (Lean.Expr.const `f [] 3992371127)
          (Lean.Expr.app
            (Lean.Expr.app
              (Lean.Expr.app (Lean.Expr.mvar `m4 9746898007) (Lean.Expr.bvar 2 3302816350191) 3308434150747)
              (Lean.Expr.bvar 1 2203303673837)
              3309994835771)
            (Lean.Expr.bvar 0 1103790997483)
            3308849464307)
          3308623335723)
        (Lean.Expr.bvar 1 2203303673837)
        3308504782939)
      2210543198593)
    1110305862629)
  9880353585
```
I think a custom written `Repr Expr` using the smart constructors will produce a more readable result. What do you think?

**Comment by Vtec234** [comment]
> There is no bootstrapping issue.

Sorry for the confusion. I was probably thinking of `Syntax` which does have this problem.

> I think a custom written `Repr Expr` using the smart constructors will produce a more readable result.

I agree. For example `Repr Expr.Data` should probably print something nicer than just the number. However even with smart constructors I would like all the excessive detail to actually be preserved. In another instance -- debugging `Syntax` -- I found myself actually wanting the detail. I was using the custom `ToFormat` [`formatStx`](https://github.com/leanprover/lean4/blob/702a24913fd25057bec78f53f363e35fec370197/src/Init/Data/Format/Syntax.lean#L41) but found that it was a bit *too* high-level and I couldn't figure out the details, so I ended up deriving `Repr` for `Syntax`. Perhaps we could have the `Repr` be very verbose still and use `ToFormat` for higher-level needs.
